### PR TITLE
[Generic DAO][Part 2] Implement backfill method

### PR DESCRIPTION
**Pull request by zhixuanjia**
## Summary
Add backfill method into Generic DAO.

## Testing Done
Unit tests

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java [hunk 1]
```diff
@@ -109,6 +136,62 @@ public Optional<GenericLocalDAO.MetadataWithExtraInfo> queryLatest(@Nonnull Urn
...
+    Map<Urn, Map<Class<? extends RecordTemplate>, Optional<? extends RecordTemplate>>> urnToAspects = new HashMap<>();
+
+    for (Urn urn : aspectClasses.keySet()) {
+      Set<Class<? extends RecordTemplate>> aspects = aspectClasses.get(urn);
+      Map<Class<? extends RecordTemplate>, Optional<? extends RecordTemplate>> aspectVals = new HashMap<>();
+
+      for (Class<? extends RecordTemplate> aspect : aspects) {
+        Optional<GenericLocalDAO.MetadataWithExtraInfo> metadata = queryLatest(urn, aspect);
+
+        metadata.ifPresent(metadataWithExtraInfo -> aspectVals.put(aspect,
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java:160]: why does value of the second inner map have to be optional? if queryLatest doesn't return anything, can we just skip putting this aspect into the map?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java [hunk 2]
```diff
@@ -109,6 +136,62 @@ public Optional<GenericLocalDAO.MetadataWithExtraInfo> queryLatest(@Nonnull Urn
...
+  /**
+   * Trigger MAE events to backfill secondary stores.
+   * @return The aspects being backfilled.
+   */
+  @Nonnull
+  public Map<Urn, Map<Class<? extends RecordTemplate>, Optional<? extends RecordTemplate>>> backfill(
+      @Nonnull BackfillMode mode, @Nonnull Map<Urn, Set<Class<? extends RecordTemplate>>> aspectClasses) {
+
+    if (aspectClasses.isEmpty()) {
+      return new HashMap<>();
```
Comment (yangyangv2) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java:148]: Collections.emptyMap()?

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanGenericLocalDAOTest.java [hunk 3]
```diff
@@ -120,12 +136,43 @@ public void testQueryLatest() throws URISyntaxException {
...
+        makeAuditStamp("tester"), null, null);
 
     Optional<GenericLocalDAO.MetadataWithExtraInfo> metadata = _genericLocalDAO.queryLatest(fooUrn, AspectFoo.class);
 
     // {"value":"bar"} is inserted later so it is the latest metadata.
     assertEquals(metadata.get().getAspect(), RecordUtils.toJsonString(aspectFoo2));
   }
+
+  @Test
+  public void testBackfill() throws URISyntaxException {
```
Comment (yangyangv2) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/EbeanGenericLocalDAOTest.java:151]: nit: adding [when / given / expect] would help for readability.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java [hunk 4]
```diff
@@ -69,20 +81,35 @@ public void setEqualityTesters(Map<Class, GenericEqualityTester> equalityTesters
...
 
       if (!latest.isPresent()) {
         saveLatest(urn, aspectClass, newValue, null, auditStamp, null);
+        _producer.produceAspectSpecificMetadataAuditEvent(urn, null, newValue, auditStamp, trackingContext, ingestionMode);
       } else {
         RecordTemplate currentValue = toRecordTemplate(aspectClass, latest.get().getAspect());
+        final AuditStamp oldAuditStamp = latest.get().getExtraInfo() == null ? null : latest.get().getExtraInfo().getAudit();
+
+        final boolean isBackfillEvent = trackingContext != null && trackingContext.hasBackfill() && trackingContext.isBackfill();
+        final boolean shouldBackfill = trackingContext != null && trackingContext.hasEmitTime() && oldAuditStamp != null && oldAuditStamp.hasTime()
```
[the remark this answers is not in the document]
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java]: agree, convert it to method and added some unit tests.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java (hunk 1 above)
Comment (jsdonn) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanGenericLocalDAO.java:160]: I missed that part thanks.

**Review by jsdonn — COMMENTED** [review]
LGTM, please get approval from Yang.

File: dao-api/src/main/java/com/linkedin/metadata/dao/producer/GenericMetadataProducer.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import javax.annotation.Nullable;
+
+
+/**
+ * Generic metadata producer without type-bound.
+ */
+public interface GenericMetadataProducer {
+
+  /**
+   * Produces an aspect specific Metadata Audit Event (MAE) after a metadata aspect is updated for an entity.
```
[the remark this answers is not in the document]
Comment (yangyangv2) [reply dao-api/src/main/java/com/linkedin/metadata/dao/producer/GenericMetadataProducer.java:18]: Thanks for clarification. I'm more from the requirement on "entity agnostic" aspect ingestion. Doesn't need to be addressed here.

**Review by yangyangv2 — APPROVED** [review]
LGTM, thanks for addressing the comments!

**Pull request merged**